skimage/util/_flood_fill_cy.pyx

```diff
@@ -0,0 +1,212 @@
+    high_tol :
+        Upper limit for tolerance comparison (ignored if `do_tol` is 0).
+    low_tol :
+        Lower limit for tolerance comparison (ignored if `do_tol` is 0).
+    """
+    cdef:
+        QueueWithHistory queue
+        unsigned char last_dim_in_neighbors
+
+    with nogil:
+        # Initialize a buffer used to queue positions while evaluating each
+        # potential maximum (flagged with 2)
+        queue_init(&queue, 64)
+        try:
+            # Conduct flood fill from this point - with or without tolerance
+            if do_tol == 0:
+                _flood_fill_do_equal(image, flags, neighbor_offsets, &queue,
+                    ravelled_seed_idx, seed_value)
+            else:
+                _flood_fill_do_tol(image, flags, neighbor_offsets, &queue,
+                    ravelled_seed_idx, seed_value, high_tol, low_tol)
+        finally:
+            # Ensure that memory is released again
+            queue_exit(&queue)
+
+
+cdef inline void _flood_fill_do_equal(
+        dtype_t[::1] image, unsigned char[::1] flags,
+        Py_ssize_t[::1] neighbor_offsets, QueueWithHistory* queue_ptr,
+        Py_ssize_t start_index, dtype_t seed_value) nogil:
+    """Fill connected areas with 1, requiring strict equality.
+
+    Parameters
+    ----------
+    image : ndarray, one-dimensional
+        The raveled view of a n-dimensional array.
+    flags : ndarray, one-dimensional
+        An array of flags that is used to store the state of each pixel during
+        evaluation.
+    neighbor_offsets : ndarray
+        A one-dimensional array that contains the offsets to find the
+        connected neighbors for any index in `image`.
+    queue_ptr :
+        Pointer to initialized queue.
+    start_index : int
+        Start position for the flood-fill.
+    seed_value :
+        Value of `image[start_index]`.
+    """
+    cdef:
+        QueueItem current_index, neighbor
+
+    # Queue start position after clearing the buffer
+    queue_clear(queue_ptr)
+    queue_push(queue_ptr, &start_index)
+    flags[start_index] = FILL
+
+    # Break loop if all queued positions were evaluated
+    while queue_pop(queue_ptr, &current_index):
+        # Look at all neighboring samples
+        for i in range(neighbor_offsets.shape[0]):
+            neighbor = current_index + neighbor_offsets[i]
+
+            # Indexing sanity check
+            if neighbor < 0:
+                continue
+            if neighbor >= image.shape[0]:
+                continue
+
+            # Shortcut if neighbor is already part of fill
+            if flags[neighbor] == UNKNOWN:
+                if image[neighbor] == seed_value:
+                # Neighbor is part of fill - but must check its neighbors too.
+                    flags[neighbor] = FILL
+                    queue_push(queue_ptr, &neighbor)
+                else:
+                    # Do not check this point again
+                    flags[neighbor] = NOT_FILL
```

Comment:
I think this may possibly be simplified further.  If checking `flags[neighbor] == UNKNOWN` has nearly the same performance as `image[neighbor] == seed_value` omitting the outer if-else-clause (with the expensive `flags[neighbor] = NOT_FILL`) may have the same or better performance and nearly the same functionality:
```python
if image[neighbor] == seed_value:
    # Neighbor is part of fill - but must check its neighbors too.
    flags[neighbor] = FILL
    queue_push(queue_ptr, &neighbor)
```

May be worth trying out and doing a quick benchmark...

Reply:
You're right, this would endlessly queue values again and again. However I still think that the last else-block with `flags[neighbor] = NOT_FILL` can be removed which should have a slight performance advantage. That point will be checked again anyway via other queued neighbors either in `== UNKOWN` or `== seed_value`.